File: 이분 탐색(결정알고리즘) & 그리디 알고리즘/마구간 정하기.py
import sys
# sys.stdin = open("input.txt", "r")

n, c = map(int, input().split())
house = [int(input()) for _ in range(n)]
house.sort()

def count(pivot):
    cnt = 0
    pre = -pivot
    for x in house:
        if x - pre >= pivot:
            cnt += 1
            pre = x
    return cnt

left = 0
right = house[-1]
result = 0
while left <= right:
    pivot = (left + right) // 2
    if count(pivot) >= c:
        left = pivot + 1
        result = pivot
    else:
        right = pivot - 1
print(result)

File: 정렬 & 이분탐색(결정알고리즘) & 투포인트 알고리즘 & 스택/마구간 정하기.cpp
#include <stdio.h>
#include <vector>
#include <algorithm>
using namespace std;

vector<int> arr;
int n;

int count(int len){
	int pos = arr[0], cnt = 1, i;
	for(i = 1; i < n; i++){
		if(arr[i] - pos >= len){
			cnt++;
			pos = arr[i];	
		}
	}
	return cnt;
}

int main(void){
	// freopen("input.txt", "r", stdin);
	int c, i, tmp, result, left, right, mid, pos, cnt;
	scanf("%d%d", &n, &c);
	
	for(i = 0; i < n; i++){
		scanf("%d", &tmp);
		arr.push_back(tmp);
	}
	sort(arr.begin(), arr.end());
	
	left = 1;
	right = arr[n - 1] - 1;
	while(left <= right){
		mid = (left + right) / 2;
		
		if(count(mid) >= c){		
			result = mid;
			left = mid + 1;	
		}
		else{
			right = mid - 1;	
		}
	}
	printf("%d", result);
	
	return 0;
}
